.licensei.toml

```diff
@@ -0,0 +1,17 @@
+approved = [
+  "apache-2.0",
+  "bsd-2-clause",
+  "bsd-3-clause",
+  "isc",
+  "mpl-2.0",
+  "mit",
+]
+
+ignored = [
+  "google.golang.org/protobuf",      # BSD-3-Clause license
+  "github.com/hashicorp/consul/api", # Mozilla Public License, version 2.0
+  "github.com/gogo/protobuf",        # FIXME: deprecated, not sure: https://github.com/gogo/protobuf/blob/master/LICENSE
+  "sigs.k8s.io/yaml",                # MIT License
+]
```

Comment:
github.com/gogo/protobuf looks like BSD-3-Clause

Reply:
At Spacelift we're using [this tool](https://github.com/google/go-licenses) internally, but it doesn't seem to be handling exceptions any better.

Ideally you should be able to ignore the license for a specific version of the package, so that if you update the package you will need to check its license again. But I'm not finding this as an option.

Reply:
Found https://github.com/goph/licensei/issues/16 :

> Currently ignoring packages is the only way to make a package pass the check without a (valid) license file.
> Unfortunately, we don't get notified about changes (ie. there is a new license).

Reply:
It is possible to add an additional configuration list "unknown" or "nolicense" and populate those that are in ignore there. Then that would solve the issue here. I've already tried that and will push a PR.

Detecting the license for consul/api specifically won't work out-of-the-box with licensei from what I saw though. The app relies on GH REST API to get the license information which uses the LICENSE file defined in the root of the repository (usually). Consul has the [BUSL](https://github.com/hashicorp/consul/blob/main/LICENSE) license in the root and [MPL2.0 under consul/api](https://github.com/hashicorp/consul/blob/main/api/LICENSE). And so to get that information, additional work would need to be implemented (perhaps using simple HTTP GET https://raw.githubusercontent.com/hashicorp/consul/main/api/LICENSE).

Reply:
Care to push this as a suggestion to this PR instead of creating a new one?

Reply:
There's a PR that has to go to licensei first and then if merged we could adopt it here (essentially replacing `ignore` with whatever the new list name would be)

Reply:
Oh ok, I didn't get that the PR you mentioned was against licensei.

Reply:
I've pushed the PR to licensei.

Reply:
Marked the PR as ready, as I believe currently we have no perfect solution, and this is probably better that nothing.

Reply:
@pdecat would you mind bunping licensei to 0.9.0 and replace `ignore` with `unknown` in the configuration? My PR is merged and released.

Reply:
> would you mind bumping licensei to 0.9.0 and replace `ignore` with `unknown` in the configuration? My PR is merged and released.

Can proceed with that, but as mentioned in Slack, I'm not sure it addresses the [concerns of licenses changes over time](https://github.com/opentffoundation/opentf/pull/310#discussion_r1317812223).

Reply:
Output after the update:
```
# make license-check
licensei check
time=2023-09-12T14:36:39.226+02:00 level=ERROR msg="github detection failed" err="no license file was found" dependency=github.com/hashicorp/consul/api
No license violations! Good job!
licensei header
```

Note: I had to delete `bin/licensei` and run `make deps` manually to get the updated version. That explains why [the documentation](https://github.com/goph/licensei#installation) recommends more complex tasks than [my original inspiration](https://github.com/bank-vaults/bank-vaults/blob/main/Makefile#L48-L50) for this.

Reply:
@marcinwyszynski @rumenvasilev @pdecat any extra discussion here?

Reply:
The ERROR message is expected, as I didn't mute that intentionally (would have been quite some deeper surgery in licensei I think), but what matters is the exit code. If you look at the way that app works now (0.9.0), you'd see that if license isn't in the whitelist and license is empty, only then unknown kicks-in. But, if the license changes from empty (unknown) to something that's not on the approved list, licensei will exit with 2 (error) failing pipeline build for example.

Reply:
> The ERROR message is expected, as I didn't mute that intentionally (would have been quite some deeper surgery in licensei I think), but what matters is the exit code. If you look at the way that app works now (0.9.0), you'd see that if license isn't in the whitelist and license is empty, only then unknown kicks-in. But, if the license changes from empty (unknown) to something that's not on the approved list, licensei will exit with 2 (error) failing pipeline build for example.

Thanks for the clarification! I do agree it does improve the situation a bit.

However, there's still the risk of a dependency going from `unknown` to `unknown`, and actually changing its license after the time it was added to the `unknown` list upon a manual license check. Which could very well happen in the case of the `github.com/hashicorp/consul/api` package moving from MPL-2.0 to something else.

We wouldn't be notified of that, would we?

Reply:
> any extra discussion here?

Just realized this PR is also missing running checks in CI.

Reply:
> Just realized this PR is also missing running checks in CI.
Sure, now CI will run :)

Reply:
I meant, currently the `licensei` command are not part of any Github workflow.

Should they be added to `checks.yml` or another one?

Reply:
> > The ERROR message is expected, as I didn't mute that intentionally (would have been quite some deeper surgery in licensei I think), but what matters is the exit code. If you look at the way that app works now (0.9.0), you'd see that if license isn't in the whitelist and license is empty, only then unknown kicks-in. But, if the license changes from empty (unknown) to something that's not on the approved list, licensei will exit with 2 (error) failing pipeline build for example.
> 
> Thanks for the clarification! I do agree it does improve the situation a bit.
> 
> However, there's still the risk of a dependency going from `unknown` to `unknown`, and actually changing its license after the time it was added to the `unknown` list upon a manual license check. Which could very well happen in the case of the `github.com/hashicorp/consul/api` package moving from MPL-2.0 to something else.
> 
> We wouldn't be notified of that, would we?

Sadly unknown to unknown won't raise any alarms, no.

In the case with hashi/consul, isn't that an isolated case? I mean, shouldn't this be solved with some additional tooling (even bash script if you will), rather than looking for the tool that supports all edge cases in the World?

In the other instances, I think the solution is rather simple. From GH's own documentation [this](https://docs.github.com/en/free-pro-team@latest/rest/licenses/licenses?apiVersion=2022-11-28#about-licenses) says:
> GitHub uses [the open source Ruby Gem Licensee](https://github.com/benbalter/licensee) to attempt to identify the license for a project. Licensee matches the contents of a project's LICENSE file (if it exists) against a short list of known licenses. As a result, the API does not take into account the licenses of project dependencies or other means of documenting a project's license such as references to the license name in the documentation.

So, any ruby enthusiasts to patch this up? - https://github.com/licensee/licensee :)

Reply:
@pdecat I think we only need two repositories in the unknown list:

```
[rumba@Rumbas-MBP opentf (main)]$ ../../public/licensei/build/licensei check 
time=2023-09-12T16:02:17.826+02:00 level=ERROR msg="github detection failed" err="no license file was found" dependency=github.com/hashicorp/consul/api
google.golang.org/protobuf: no license file found
github.com/hashicorp/consul/api: no license file found
```
The other two are detected correctly.

Reply:
I may have found an alternative option running `go mod vendor` before running `licensei check` without any use of `ignore` or `unknown` options as licensei takes these into account: https://github.com/goph/licensei/blob/v0.9.0/internal/licensei/license.go#L58

Version pinning is then ensured by `go.mod`/`go.sum`.

Reply:
Hmm, now, there's only one dependency causing issue which actually has no license at all:

```
# make license-check
go mod vendor
licensei check
time=2023-09-12T17:50:23.828+02:00 level=ERROR msg="github detection failed" err="no license file was found" dependency=github.com/hashicorp/consul/api
time=2023-09-12T17:50:31.504+02:00 level=ERROR msg="sourced detection failed" err="no license file was found" dependency=github.com/aliyun/aliyun-oss-go-sdk
github.com/aliyun/aliyun-oss-go-sdk: no license file found
make: *** [Makefile:59: license-check] Error 2
```

But it has an MIT license: https://github.com/aliyun/aliyun-oss-go-sdk/blob/master/LICENSE

Reply:
Updating this dependency makes thing pass:
```
# make license-check
go mod vendor
licensei check
time=2023-09-12T17:56:53.343+02:00 level=ERROR msg="github detection failed" err="no license file was found" dependency=github.com/hashicorp/consul/api
No license violations! Good job!
licensei header

```
License was apparently added afterward.

Reply:
Follow up on https://github.com/opentffoundation/opentf/pull/310#discussion_r1323137019:

Downgrading licensei from 0.9.0 to 0.8.0 brings back the two issues with github.com/gogo/protobuf and sigs.k8s.io/yaml:

```
# sed -i s/0.9.0/0.8.0/ Makefile

# make deps
curl -sfL https://git.io/licensei | bash -s v0.8.0
goph/licensei info checking GitHub for tag 'v0.8.0'
goph/licensei info found version: 0.8.0 for v0.8.0/linux/amd64
goph/licensei info installed ./bin/licensei

# make license-check
go mod vendor
licensei check
time=2023-09-12T18:13:22.147+02:00 level=ERROR msg="github detection failed" err="no license file was found" dependency=github.com/hashicorp/consul/api
time=2023-09-12T18:13:23.271+02:00 level=ERROR msg="sourced detection failed" err="no license file was found" dependency=github.com/gogo/protobuf
time=2023-09-12T18:13:26.629+02:00 level=ERROR msg="sourced detection failed" err="no license file was found" dependency=sigs.k8s.io/yaml
github.com/gogo/protobuf: no license file found
sigs.k8s.io/yaml: no license file found
make: *** [Makefile:59: license-check] Error 2
```

Something strange changed in `licensei` behavior between those versions...

Reply:
Maybe it is the result of this dependency update: 

https://github.com/goph/licensei/pull/57
https://github.com/goph/licensei/commit/08b93420a16429f02dfc647dc602aaa00d3532da
https://github.com/go-enry/go-license-detector/compare/v4.3.0...v4.3.1

Reply:
Hey all
This discussion is very long and hard to follow. Are there any action items here to fix?

Reply:
I believe all issue are now resolved.